invenio_app_ils/records/permissions.py

```diff
@@ -0,0 +1,164 @@
+"""Ils records' permissions."""
+from flask_login import current_user
+from flask_principal import RoleNeed
+from invenio_access import DynamicPermission, action_factory
+from invenio_admin.permissions import action_admin_access
+
+from invenio_app_ils.records.utils import get_user_provides
+
+librarian_role = RoleNeed('librarian')
+create_action = action_factory('create')
+
+
+def record_permission_factory(record=None, action=None):
+    """Record permission factory."""
+    return RecordPermission.create(record, action)
+
+
+def record_create_permission_factory(record=None):
+    """Create permission factory."""
+    return record_permission_factory(record=record, action='create')
+
+
+def record_read_permission_factory(record=None):
+    """Read permission factory."""
+    return record_permission_factory(record=record, action='read')
+
+
+def record_update_permission_factory(record=None):
+    """Update permission factory."""
+    return record_permission_factory(record=record, action='update')
+
+
+def record_delete_permission_factory(record=None):
+    """Delete permission factory."""
+    return record_permission_factory(record=record, action='delete')
+
+
+class RecordPermission(object):
+    """Record permission.
+
+    - Create action given to librarian, admin and specified users.
+    - Read access given to everyone with possibility to hide.
+    - Delete access to admin and specified users.
+    """
+
+    create_actions = ['create']
+    read_actions = ['read']
+    update_actions = ['update']
+    delete_actions = ['delete']
+
+    def __init__(self, record, func, user):
+        """Initialize a file permission object."""
+        self.record = record
+        self.func = func
+        self.user = user or current_user
+
+    def can(self):
+        """Determine access."""
+        return self.func(self.user, self.record)
+
+    @classmethod
+    def create(cls, record, action, user=None):
+        """Create a record permission."""
+        if action in cls.read_actions:
+            return cls(record, has_read_record_permission, user)
+        elif action in cls.create_actions:
+            return cls(record, has_create_record_permission, user)
+        elif action in cls.update_actions:
+            return cls(record, has_update_record_permission, user)
+        elif action in cls.delete_actions:
+            return cls(record, has_delete_record_permission, user)
+        else:
+            return cls(record, deny, user)
+
+
+#
+# Utility functions
+#
+def deny(user, record):
+    """Deny access."""
+    return False
+
+
+def allow(user, record):
+    """Allow access."""
+    return True
+
+
+def is_public(data, action):
+    """Check if the record is fully public.
+
+    In practice this means that the record doesn't have the ``access`` key or
+    the action is not inside access or is empty.
+    """
+    return '_access' not in data or not data.get('_access', {}).get(action)
+
+
+def has_record_permission(user, record, action):
```

Comment:
it might be possible to take full advantage of `invenio-access` and rewrite this part as (pseudo-code here, warning!):
```
class RecordPermission(Permission):
    """Return Permission to evaluate if the current user can perform the action in the given record."""
    
    def __init__(self, record, action):
        """Constructor."""
        allowed_users = record.get('_access', {}).get(action, [])
        user_needs = [UserNeed(user_id) for user_id in allowed_users]
        super(RecordPermission, self).__init__(**user_needs)
```

And you don't need the utils.
The point is to write to minimize the code we have to create/maintain, and rely on the invenio packages.
**warning**: maybe this doesn't work :(